cmd/lotus-shed/indexes.go

```diff
@@ -879,3 +900,339 @@ var backfillTxHashCmd = &cli.Command{
+		if err := tx.Rollback(); err != nil {
+			fmt.Printf("ERROR: rollback %s tx :%s", tableName, err)
+		}
+	}
+
+	res, err := tx.Exec(deleteEventFromStartHeight, startHeight)
+	if err != nil {
+		rollback(eventTable)
+		return xerrors.Errorf("error deleting event: %w", err)
+	}
+
+	var totalRowsAffected int64
+	rowsAffected, err := res.RowsAffected()
+	if err != nil {
+		return xerrors.Errorf("error getting rows affected: %w", err)
+	}
+
+	totalRowsAffected += rowsAffected
+	log.Infof("pruned %s table, rows affected: %d", eventTable, rowsAffected)
+
+	res, err = tx.Exec(deleteEventEntriesByEventIDs, startHeight)
+	if err != nil {
+		rollback(eventEntryTable)
+		return xerrors.Errorf("error deleting event entries: %w", err)
+	}
+
+	rowsAffected, err = res.RowsAffected()
+	if err != nil {
+		return xerrors.Errorf("error getting rows affected: %w", err)
+	}
+
+	totalRowsAffected += rowsAffected
+	log.Infof("pruned %s table, rows affected: %d", eventEntryTable, rowsAffected)
+
+	res, err = tx.Exec(deleteEventsSeenByHeight, startHeight)
+	if err != nil {
+		rollback(eventsSeenTable)
+		return xerrors.Errorf("error deleting events seen: %w", err)
+	}
+
+	rowsAffected, err = res.RowsAffected()
+	if err != nil {
+		return xerrors.Errorf("error getting rows affected: %w", err)
+	}
+
+	err = tx.Commit()
+	if err != nil {
+		return xerrors.Errorf("error committing tx: %w", err)
+	}
+
+	totalRowsAffected += rowsAffected
+	log.Infof("pruned %s table, rows affected: %d", eventsSeenTable, rowsAffected)
+
+	log.Infof("Done pruning all events indexes, total rows affected: %d", totalRowsAffected)
+	return nil
+}
+
+// pruneTxIndex is a helper function that prunes the txindex.db for a number of epochs starting from a specified height
+func pruneTxIndex(ctx context.Context, api lapi.FullNode, basePath string, startHeight int64) error {
+	log.Infof("pruning txindex")
+
+	dbPath := path.Join(basePath, databaseType, ethhashlookup.DefaultDbFilename)
+	db, err := sql.Open(databaseDriver, dbPath)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		err := db.Close()
+		if err != nil {
+			fmt.Printf("ERROR: closing db: %s", err)
+		}
+	}()
+
+	deleteStmt, err := db.Prepare(deleteTxHashIndexByCID)
+	if err != nil {
+		return xerrors.Errorf("failed to prepare tx hash index delete statement: %w", err)
+	}
+
+	// we want to start pruning older than the specified height
+	startHeight = startHeight - 1
+
+	currTs, err := api.ChainGetTipSetByHeight(ctx, abi.ChainEpoch(startHeight), types.EmptyTSK)
+	if err != nil {
+		return err
+	}
+
+	var totalRowsAffected int64
+	for currTs != nil {
+		select {
+		case <-ctx.Done():
+			fmt.Println("request cancelled")
+			return nil
+		default:
+		}
+
+		for _, blockheader := range currTs.Blocks() {
+			blkMsgs, err := api.ChainGetBlockMessages(ctx, blockheader.Cid())
+			if err != nil {
+				log.Infof("failed to get block messages at height: %d", currTs.Height())
+				continue // should we break here?
```

Comment:
@rvagg should we break here or keep going till the first tipset?

Is there a possibility, that a block is missing in between?

While testing I saw one block was missing, after that, it kept going on like this.


Testing logs:
`2024-08-21T20:21:39.118+0530    INFO    lotus-shed      lotus-shed/indexes.go:1148      failed to get block messages at height: 1867646

2024-08-21T20:21:39.122+0530    INFO    lotus-shed      lotus-shed/indexes.go:1148      failed to get block messages at height: 1867641

2024-08-21T20:21:39.127+0530    INFO    lotus-shed      lotus-shed/indexes.go:1148      failed to get block messages at height: 1867633

2024-08-21T20:21:39.134+0530    INFO    lotus-shed      lotus-shed/indexes.go:1148      failed to get block messages at height: 1867623`

Reply:
Oh, this one is tricky! because we don't have an epoch to delete against we can't do it bulk, as you've discovered, but we also can't do it this way because we kind of want the inverse: we want to delete entries for messages where we _don't_ have the message in the blockstore ourselves. So it's almost like: collect all messages, then delete everything not in that list. Which is impractical because that could get very large. The other problem with using the chain to do this is that there are going to be messages not on the canonical chain that we need to consider (i.e. there's been a reorg, so walking back from some tip doesn't lead us to tipsets that were previously considered canonical but are no longer), perhaps we want to include them, perhaps we should delete them because they are non-canonical, it's actually not even clear what you'd want to do because it depends on how the lookup API is being used. But conceptually what you'd want is something like: nothing in this txhashdb that isn't also in my blockstore. Because if I can't get to the original message anyway then there's no real help in being able to translate it from an eth hash.

BUT; just doing a quick look at the db, it's the one that we already have GC in already, and we achieve that by timestamp. If you turn on `EthTxHashMappingLifetimeDays` then it'll clean up according to the timestamp it was inserted at.

So I think the strategy needs to change for this table: we just can't do it on epoch unfortunately; we either need to not offer it here at all, or we need to come up with a timestamp and run something like the existing functionality: https://github.com/filecoin-project/lotus/blob/dbef5ded1585bcdcf84bbe09f3908210f9775ef6/chain/ethhashlookup/eth_transaction_hash_lookup.go#L131

We _can_ get a timestamp. So, if you want to keep this pruning here, you could keep the existing `api.ChainGetTipSetByHeight(ctx, abi.ChainEpoch(startHeight), types.EmptyTSK)` to get the tipset, look at the first block and get its timestamp (e.g. [like this](https://github.com/filecoin-project/lotus/blob/dbef5ded1585bcdcf84bbe09f3908210f9775ef6/blockstore/splitstore/splitstore_compact.go#L131) and then work out how many seconds ago it is and then you can run the delete using `datetime()` maths. Something like `DELETE FROM eth_tx_hashes WHERE insertion_time < datetime('now', '-X seconds')`. Note how sqlite datetime maths works:

```
sqlite> select datetime('now'), datetime('now', '-10 seconds'), datetime('now', '-86400 seconds');
2024-08-22 03:38:03|2024-08-22 03:37:53|2024-08-21 03:38:03
```

This is one area where we could apply some reuse of existing code: in `eth_transaction_hash_lookup.go`, we could expose a method to reuse the existing delete statement, and since we only run this once per hour on the existing GC we can get rid of the preparedstatement too and expose a public function to run it on any instance of a db, here's a quick go at what what may look like over in that file:

```go
func (ei *EthTxHashLookup) DeleteEntriesOlderThan(days int) (int64, error) {
	if ei.db == nil {
		return 0, xerrors.New("db closed")
	}
	epochs := abi.ChainEpoch(build.BlockDelaySecs * builtin.SecondsInDay * uint64(days))
	return DeleteEntriesOlderThan(ei.db, epochs)
}

// DeleteEntriesOlderThan deletes entries older than the given number of epochs from now.
func DeleteEntriesOlderThan(db *sql.DB, epochs abi.ChainEpoch) (int64, error) {
	secondsAgo := int64(epochs) * int64(build.BlockDelaySecs)
	res, err := db.Exec(deleteOlderThan, "-"+strconv.FormatInt(secondsAgo, 10)+" seconds")
	if err != nil {
		return 0, err
	}
	return res.RowsAffected()
}
```

Reply:
Updated as per your suggestion of reusing code.